packages/compiler/src/core/name-resolver.ts

```diff
@@ -0,0 +1,1297 @@
+      // For alias if the alias itself is a template declaration then its not actually instantiating the reference
+      const isTemplateInstantiation =
+        aliasResult.isTemplateInstantiation && aliasNode.templateParameters.length === 0;
+      if (isTemplateInstantiation) {
+        links.isTemplateInstantiation = true;
+      }
+      if (aliasResult.finalSymbol) {
+        links.finalSymbol = aliasResult.finalSymbol;
+      }
+      result = {
+        ...aliasResult,
+        finalSymbol: links.finalSymbol,
+        isTemplateInstantiation: result.isTemplateInstantiation || isTemplateInstantiation,
+      };
+    } else if (resolvedSym && resolvedSym.flags & SymbolFlags.TemplateParameter) {
+      // references to template parameters with constraints can reference the
+      // constraint type members
+      const templateNode = resolvedSym.declarations[0] as TemplateParameterDeclarationNode;
+      if (templateNode.constraint) {
+        result = resolveTemplateParameter(templateNode);
+      }
+    }
+
+    // make sure we've bound and fully resolved the referenced
+    // node before returning it.
+    if (resolvedSym) {
+      if (
+        resolvedSym.flags & SymbolFlags.Declaration &&
+        ~resolvedSym.flags & SymbolFlags.Namespace // TODO: check, this breaks usings if we don't have that as we start checking before the usings have been resolved.
+      ) {
+        bindAndResolveNode(resolvedSym.declarations[0]);
+      }
+      // TODO: non-declarations - this doesn't seem needed
+    }
+
+    return result;
+  }
+
+  function resolveTypeReferenceWorker(
+    node: TypeReferenceNode | MemberExpressionNode | IdentifierNode,
+    options: ResolveTypReferenceOptions,
+  ): ResolutionResult {
+    if (node.kind === SyntaxKind.TypeReference) {
+      const result = resolveTypeReference(node.target, options);
+      return node.arguments.length > 0 ? { ...result, isTemplateInstantiation: true } : result;
+    } else if (node.kind === SyntaxKind.MemberExpression) {
+      return resolveMemberExpression(node, options);
+    } else if (node.kind === SyntaxKind.Identifier) {
+      return resolveIdentifier(node, options);
+    }
+
+    compilerAssert(false, "Unexpected node kind");
+  }
+
+  function resolveMemberExpression(
+    node: MemberExpressionNode,
+    options: ResolveTypReferenceOptions,
+  ): ResolutionResult {
+    const baseResult = resolveTypeReference(node.base, {
+      ...options,
+      resolveDecorators: false, // When resolving the base it can never be a decorator
+    });
+
+    if (baseResult.resolutionResult & ResolutionResultFlags.ResolutionFailed) {
+      return baseResult;
+    }
+    const baseSym = baseResult.finalSymbol;
+
+    compilerAssert(baseSym, "Base symbol must be defined if resolution did not fail");
+    const memberResult = resolveMemberExpressionForSym(baseSym, node, options);
+
+    // TODO: is it better to do this or just when you resolve identifier have to maybe check up the parent once.
```

Comment:
No strong opinion here.

Reply:
this ends up being quite nice so you don't have to care about where you are the links always tell you what it links to